### PR TITLE
Bound staking storage items

### PR DESCRIPTION
**Pull request by Ank4n**
This fixes part 1 & 2 of https://github.com/paritytech/substrate/issues/12187 

polkadot companion: https://github.com/paritytech/polkadot/pull/5996

### CODE REVIEW

File: frame/staking/src/migrations.rs [hunk 1]
```diff
@@ -20,6 +20,36 @@ use super::*;
 use frame_election_provider_support::SortedListProvider;
 use frame_support::traits::OnRuntimeUpgrade;
 
+pub mod v11 {
```
Comment (kianenigma) [inline frame/staking/src/migrations.rs:23]: @ruseinov given you past experience, would like your review of this :)

File: frame/staking/src/pallet/impls.rs [hunk 2]
```diff
@@ -118,15 +118,17 @@ impl<T: Config> Pallet<T> {
...
 		ledger
 			.claimed_rewards
 			.retain(|&x| x >= current_era.saturating_sub(history_depth));
+			
 		match ledger.claimed_rewards.binary_search(&era) {
 			Ok(_) =>
 				return Err(Error::<T>::AlreadyClaimed
 					.with_weight(T::WeightInfo::payout_stakers_alive_staked(0))),
-			Err(pos) => ledger.claimed_rewards.insert(pos, era),
+			Err(pos) => ledger.claimed_rewards.try_insert(pos, era)
```
Comment (kianenigma) [inline frame/staking/src/pallet/impls.rs:130]: some explanation of why this is defensive would always be good.

File: frame/staking/src/pallet/impls.rs [hunk 3]
```diff
@@ -870,6 +872,15 @@ impl<T: Config> Pallet<T> {
 			DispatchClass::Mandatory,
 		);
 	}
+
+	/// This will return the currently configured History Depth
```
Comment (kianenigma) [inline frame/staking/src/pallet/impls.rs:876]: my OCD-comments 😁: 

1. punctuation missing. 
2. line wrapping should ideally be 100, not less :p

File: frame/staking/src/pallet/mod.rs [hunk 4]
```diff
@@ -124,6 +125,11 @@ pub mod pallet {
 		#[pallet::constant]
 		type MaxNominations: Get<u32>;
 
+		/// Era History Depth for the validator rewards to be claimed. 
+		/// This should never be decreased once set.
```
Comment (kianenigma) [inline frame/staking/src/pallet/mod.rs:129]: We can probably do a better doc here.

File: frame/staking/src/pallet/mod.rs [hunk 5]
```diff
@@ -124,6 +125,18 @@ pub mod pallet {
 		#[pallet::constant]
 		type MaxNominations: Get<u32>;
 
+		/// `HistoryDepth` stores the number of previous eras for which
```
Comment (kianenigma) [inline frame/staking/src/pallet/mod.rs:128]: I think this value is actually used for the first key of a number of double maps as well 🙈 

(also, same comment about comments being 100 in width, I used this https://marketplace.visualstudio.com/items?itemName=stkb.rewrap)

File: frame/staking/src/pallet/mod.rs [hunk 6]
```diff
@@ -802,7 +806,7 @@ pub mod pallet {
 			<Payee<T>>::insert(&stash, payee);
 
 			let current_era = CurrentEra::<T>::get().unwrap_or(0);
-			let history_depth = Self::history_depth();
+			let history_depth = T::HistoryDepth::get();
```
Comment (kianenigma) [inline frame/staking/src/pallet/mod.rs:809]: This would mean that we should re-benchmark, and expect one less read now with this PR.

**Comment by Ank4n** [comment]
/cmd queue -c bench-bot $ pallet dev pallet_staking

File: frame/staking/src/pallet/mod.rs (hunk 4 above)
Comment (Ank4n) [reply frame/staking/src/pallet/mod.rs]: I added more doc based on your feedback. Have a look again.

File: frame/staking/src/pallet/mod.rs (hunk 5 above)
Comment (Ank4n) [reply frame/staking/src/pallet/mod.rs]: I could see this used only in Claimed Rewards in the Staking Ledger

**Comment by Ank4n** [comment]
/cmd queue -c bench-bot $ pallet dev pallet_staking

File: frame/staking/src/migrations.rs [hunk 7]
```diff
@@ -20,6 +20,36 @@ use super::*;
...
+
+	/// Clean up `HistoryDepth` from storage
+	///
+	/// We will be depending on the configurable value of `HistoryDepth` post
+	/// this release.
+	pub struct MigrateToV11<T>(sp_std::marker::PhantomData<T>);
+	impl<T: Config> OnRuntimeUpgrade for MigrateToV11<T> {
+		fn on_runtime_upgrade() -> frame_support::weights::Weight {
+			if StorageVersion::<T>::get() == Releases::V10_0_0 {
+				// may be just check its greater?
```
Comment (kianenigma) [inline frame/staking/src/migrations.rs:38]:
```suggestion
```

File: frame/staking/src/migrations.rs [hunk 8]
```diff
@@ -20,6 +20,36 @@ use super::*;
...
+				debug_assert_eq!(T::HistoryDepth::get(), HistoryDepth::<T>::get());
```
Comment (kianenigma) [inline frame/staking/src/migrations.rs:39]: should go in `pre_migrate`

File: frame/staking/src/migrations.rs [hunk 9]
```diff
@@ -20,6 +20,36 @@ use super::*;
...
+				HistoryDepth::<T>::kill();
+				StorageVersion::<T>::put(Releases::V11_0_0);
+
+				log!(info, "MigrateToV11 executed successfully");
+				T::DbWeight::get().reads_writes(1, 1)
```
Comment (kianenigma) [inline frame/staking/src/migrations.rs:44]:
```suggestion
				T::DbWeight::get().reads_writes(2, 2)
```

File: frame/staking/src/pallet/mod.rs [hunk 10]
```diff
@@ -124,6 +125,19 @@ pub mod pallet {
...
+		/// `HistoryDepth` is the number of eras to keep in history.
+		///
+		/// This used to be a storage value previously. If you are migrating
+		/// from storage value to config value, you should read this value from
+		/// storage and set the same value in configuration.
+		///
+		/// Note: `HistoryDepth` is used as the upper bound for the `BoundedVec`
+		/// item `StakingLedger.claimed_rewards`. Setting this value lower than
+		/// the existing value can lead to inconsistencies and will need to be
+		/// handled properly in migration.
```
Comment (kianenigma) [inline frame/staking/src/pallet/mod.rs:137]:
```suggestion
		/// handled properly in a migration.
```

File: frame/staking/src/pallet/mod.rs [hunk 11]
```diff
@@ -202,22 +216,6 @@ pub mod pallet {
...
 	}
 
-	#[pallet::type_value]
-	pub(crate) fn HistoryDepthOnEmpty() -> u32 {
-		84u32
-	}
-
-	/// Number of eras to keep in history.
-	///
-	/// Information is kept for eras in `[current_era - history_depth; current_era]`.
```
Comment (kianenigma) [inline frame/staking/src/pallet/mod.rs:212]: Something similar to this document should be added to the new type config.

File: frame/staking/src/pallet/mod.rs [hunk 12]
```diff
@@ -813,7 +818,12 @@ pub mod pallet {
 				total: value,
 				active: value,
 				unlocking: Default::default(),
-				claimed_rewards: (last_reward_era..current_era).collect(),
+				claimed_rewards: (last_reward_era..current_era)
```
Comment (kianenigma) [inline frame/staking/src/pallet/mod.rs:821]: would be nice to make sure this is checked with some reasonable values in tests that test the `bond` transaction. Good reason to play around with a few of the tests as well.

**Review by kianenigma — APPROVED** [review]
Already at a good shape, but could use a little bit more work. 

🎉

File: frame/staking/src/migrations.rs (hunk 8 above)
Comment (Ank4n) [reply frame/staking/src/migrations.rs:39]: @kianenigma Do you mean we should fail migration if this condition does not match? Or move as it is (debug assertion) in the pre hook?

File: frame/staking/src/migrations.rs (hunk 9 above)
Comment (Ank4n) [reply frame/staking/src/migrations.rs:44]: I assume this is because there are 2 reads and 2 writes in the migration? Seems to me though some other migrations do not have the correct numbers that I would expect. Such as in v10, I think there is atleast 2 reads and 2 writes, in addition to call to `apply_slash`. Is that a mistake?

File: frame/staking/src/pallet/mod.rs (hunk 12 above)
Comment (Ank4n) [reply frame/staking/src/pallet/mod.rs:827]: Some tests were already testing these values. I refactored them a bit to read/calculate first and last reward era in code. I also changed HistoryDepth value in staking mock runtime to 80 to keep it a bit different from actual runtime and make sure in test we are not still reading the hardcoded value `84` like earlier.

Comment (Ank4n) [reply frame/staking/src/pallet/mod.rs:827]: Actually none of them is testing the claimed rewards after a bond transaction. I will add that.

**Comment by Ank4n** [comment]
/cmd queue -c bench-bot $ pallet dev pallet_staking

**Review by ruseinov — APPROVED** [review]
seems fine by me.

It would also be nice to try the migration out using try-runtime before merging this. This could be a bit tricky due to the fact that you also need all the previous migrations to have been executed beforehand. AFAIK when you try executing several migrations with try-runtime on the same snapshot - it will fail due to pre_upgrade checks being executed before the actual migrations. 

One way to it I guess is to run a local chain synced and with all the migrations ran up to yours.

**Comment by ruseinov** [comment]
bot rebase

**Comment by Ank4n** [comment]
bot rebase

File: frame/staking/src/lib.rs [hunk 13]
```diff
@@ -887,6 +898,7 @@ enum Releases {
 	V9_0_0,  // inject validators into `VoterList` as well.
 	V10_0_0, // remove `EarliestUnappliedSlash`.
 	V11_0_0, // Move pallet storage prefix, e.g. BagsList -> VoterBagsList
+	V12_0_0, // remove `HistoryDepth`
```
Comment (kianenigma) [inline frame/staking/src/lib.rs:901]:
```suggestion
	V12_0_0, // remove `HistoryDepth`.
```

File: frame/staking/src/migrations.rs [hunk 14]
```diff
@@ -20,6 +20,58 @@ use super::*;
...
+	use super::*;
+	use frame_support::{pallet_prelude::ValueQuery, storage_alias};
+
+	#[storage_alias]
+	type HistoryDepth<T: Config> = StorageValue<Pallet<T>, u32, ValueQuery>;
+
+	/// Clean up `HistoryDepth` from storage
+	///
+	/// We will be depending on the configurable value of `HistoryDepth` post
+	/// this release.
```
Comment (kianenigma) [inline frame/staking/src/migrations.rs:33]:
```suggestion
	/// Clean up `HistoryDepth` from storage.
	///
	/// We will be depending on the configurable value of `HistoryDepth` post
	/// this release.
```

File: frame/staking/src/pallet/mod.rs [hunk 15]
```diff
@@ -124,6 +125,25 @@ pub mod pallet {
 		#[pallet::constant]
 		type MaxNominations: Get<u32>;
 
+		/// Number of eras to keep in history.
+		///
+		/// Information is kept for eras in `[current_era - history_depth;
+		/// current_era]`.
+		///
+		/// Must be more than the number of eras delayed by session otherwise.
```
Comment (kianenigma) [inline frame/staking/src/pallet/mod.rs:133]:
```suggestion
		/// Must be more than the number of eras delayed by session.
```

File: frame/staking/src/pallet/mod.rs [hunk 16]
```diff
@@ -124,6 +125,25 @@ pub mod pallet {
 		#[pallet::constant]
 		type MaxNominations: Get<u32>;
 
+		/// Number of eras to keep in history.
+		///
+		/// Information is kept for eras in `[current_era - history_depth;
```
Comment (kianenigma) [inline frame/staking/src/pallet/mod.rs:130]: what is "information"?

answer: `Ledger.claimed_rewards`, and a number of era information such as `ErasStakers`, as we discussed.

File: frame/staking/src/tests.rs [hunk 17]
```diff
@@ -5287,3 +5292,160 @@ fn proportional_ledger_slash_works() {
...
+		assert_ok!(Staking::bond(Origin::signed(3), 4, 1500, RewardDestination::Controller));
+
+		// all previous era before the bonding action should be marked as
+		// claimed.
+		let mut claimed_rewards = vec![];
+		for i in start_reward_era..=last_reward_era {
+			claimed_rewards.push(i);
+		}
+
+		let claimed_rewards: BoundedVec<_, _> = claimed_rewards.try_into().unwrap();
```
Comment (kianenigma) [inline frame/staking/src/tests.rs:5323]:
```suggestion
		
		let claimed_rewards: BoundedVec<_, _> = (start_reward_era..=last_reward_era).collect::<Vec<_>>().try_into().unwrap();
```

File: frame/staking/src/tests.rs [hunk 18]
```diff
@@ -5287,3 +5292,160 @@ fn proportional_ledger_slash_works() {
...
+
+		// start next era
+		current_era = current_era + 1;
+		mock::start_active_era(current_era);
+
+		// claiming reward for last era in which validator was active works
+		assert_ok!(Staking::payout_stakers(Origin::signed(4), 3, current_era - 1));
+
+		// consumed weight for all payout_stakers dispatches that fail
+		let err_weight = weights::SubstrateWeight::<Test>::payout_stakers_alive_staked(0);
```
Comment (kianenigma) [inline frame/staking/src/tests.rs:5343]:
```suggestion
		let err_weight = weights::SubstrateWeight::<Test>::payout_stakers_alive_staked(0);
		
```

no no -- from within the pallet, we only know `<Test/T as Config>::WeightInfo`, should not assume that it is `SubstrateWeight`

File: frame/staking/src/tests.rs [hunk 19]
```diff
@@ -5287,3 +5292,160 @@ fn proportional_ledger_slash_works() {
...
+
+		// start current era
+		mock::start_active_era(current_era);
+
+		// add a new candidate for being a staker. account 3 controlled by 4.
+		assert_ok!(Staking::bond(Origin::signed(3), 4, 1500, RewardDestination::Controller));
+
+		// all previous era before the bonding action should be marked as
+		// claimed.
+		let mut claimed_rewards = vec![];
```
Comment (kianenigma) [inline frame/staking/src/tests.rs:5388]: same as above

File: frame/staking/src/tests.rs [hunk 20]
```diff
@@ -5287,3 +5292,160 @@ fn proportional_ledger_slash_works() {
...
+		for i in start_reward_era..=last_reward_era {
+			claimed_rewards.push(i);
+		}
+
+		let claimed_rewards: BoundedVec<_, _> = claimed_rewards.try_into().unwrap();
+		assert_eq!(
```
Comment (kianenigma) [inline frame/staking/src/tests.rs:5394]: how is this assertion actually passing? are we claiming rewards in `start_active_era`?

File: frame/staking/src/tests.rs [hunk 21]
```diff
@@ -5287,3 +5292,160 @@ fn proportional_ledger_slash_works() {
 		BTreeMap::from([(4, 0), (5, value_slashed), (6, 0), (7, 0)])
 	);
 }
+
+#[test]
+fn pre_bonding_era_cannot_be_claimed() {
+	// Verifies initial conditions of mock
```
Comment (kianenigma) [inline frame/staking/src/tests.rs:5298]: I am not following it strictly myself, but I suggest this about how to document/structure tests: 

https://www.agilealliance.org/glossary/gwt/#:~:text=An%20example%3A,complete%20without%20errors%20or%20warnings

File: frame/staking/src/tests.rs [hunk 22]
```diff
@@ -5287,3 +5292,160 @@ fn proportional_ledger_slash_works() {
...
+		// claiming reward for last era in which validator was active works
+		assert_ok!(Staking::payout_stakers(Origin::signed(4), 3, current_era - 1));
+
+		// next era
+		current_era = current_era + 1;
+		mock::start_active_era(current_era);
+
+		// history_depth reduced without migration
+		let history_depth = original_history_depth - 1;
+		HistoryDepth::set(history_depth);
```
Comment (kianenigma) [inline frame/staking/src/tests.rs:5418]:
```suggestion
		HistoryDepth::mutate(|x| *x -= 1);
```

I think we also implement this.

File: frame/staking/src/tests.rs [hunk 23]
```diff
@@ -5287,3 +5292,160 @@ fn proportional_ledger_slash_works() {
...
+		// claiming reward does not work anymore
+		assert_noop!(
+			Staking::payout_stakers(Origin::signed(4), 3, current_era - 1),
+			Error::<Test>::NotController
+		);
+
+		// new stakers can still bond
+		// add new staker works
```
Comment (kianenigma) [inline frame/staking/src/tests.rs:5426]:
```suggestion
```

File: frame/staking/src/pallet/mod.rs (hunk 16 above)
Comment (Ank4n) [reply frame/staking/src/pallet/mod.rs]: https://github.com/paritytech/substrate/pull/12230/commits/121557df70409447c70c548256e15fe4626a3f6e

File: frame/staking/src/tests.rs (hunk 20 above)
Comment (Ank4n) [reply frame/staking/src/tests.rs:5384]: No. Its like a little trick from what I understand. When a staker bonds for the first time, all the eras before it are marked as rewards already claimed. This means when a staker tries to claim reward for an era before they started staking, they are not able to since its pre-marked as claimed.

File: frame/staking/src/tests.rs [hunk 24]
```diff
@@ -5287,3 +5292,160 @@ fn proportional_ledger_slash_works() {
...
+		// consumed weight for all payout_stakers dispatches that fail
+		let err_weight = weights::SubstrateWeight::<Test>::payout_stakers_alive_staked(0);
+		// cannot claim rewards for an era before bonding occured as it is
+		// already marked as claimed.
+		assert_noop!(
+			Staking::payout_stakers(Origin::signed(4), 3, current_era - 2),
+			Error::<Test>::AlreadyClaimed.with_weight(err_weight)
+		);
+
+		// decoding will fail now since Staking Ledger is in corrupt state
```
[the remark this answers is not in the document]
Comment (Ank4n) [reply frame/staking/src/tests.rs:5345]: I was thinking that too but I left it since I wanted the test to show that invalid era rewards still cannot be claimed even if the StakingLedger is in a corrupt state. Technically this is tested in the next test already since in reality, even the valid era cannot be claimed. I could go either way though and remove it.

File: frame/staking/src/tests.rs (hunk 22 above)
Comment (Ank4n) [reply frame/staking/src/tests.rs:5408]: I need to keep track of `history_depth` (current) for the assertions after this and `original_history_depth` to fix the corrupt state in order for the `post_conditions()` of the test to pass.

File: frame/staking/src/tests.rs (hunk 21 above)
Comment (Ank4n) [reply frame/staking/src/tests.rs:5298]: Its a little bit difficult with such long tests. Right now its a bit more like given, when1, then1, when2, then2. To strictly follow this we should have every test asserting on only one set of actions (then) and repeat the then1 actions in the `given` for the following test.

File: frame/staking/src/tests.rs (hunk 18 above)
Comment (Ank4n) [reply frame/staking/src/tests.rs]: https://github.com/paritytech/substrate/pull/12230/commits/a8f0e7df76f6ac83e9424f52b176282b3468d800#diff-64c049b4e94a55bdeaf757c725cf7d1df623b754557af4ced157024a943c4703R3311

File: frame/staking/src/tests.rs (hunk 21 above)
Comment (kianenigma) [reply frame/staking/src/tests.rs:5298]: consider it a suggestion for now then. Yeah, kinda hard to follow in any no-trivial test.

File: frame/staking/src/tests.rs (hunk 18 above)
Comment (kianenigma) [reply frame/staking/src/tests.rs]: huh, both are wrong and should be fixed then I presume.

**Comment by Ank4n** [comment]
bot rebase

**Comment by Ank4n** [comment]
bot merge

File: frame/staking/src/weights.rs [hunk 25]
```diff
@@ -372,52 +367,54 @@ impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
...
+	/// The range of component `v` is `[500, 1000]`.
+	/// The range of component `n` is `[500, 1000]`.
+	/// The range of component `s` is `[1, 20]`.
+	fn get_npos_voters(v: u32, n: u32, _s: u32, ) -> Weight {
+		Weight::from_ref_time(24_930_788_000 as u64)
+			// Standard Error: 266_386
+			.saturating_add(Weight::from_ref_time(6_687_552 as u64).saturating_mul(v as u64))
+			// Standard Error: 266_386
+			.saturating_add(Weight::from_ref_time(6_839_134 as u64).saturating_mul(n as u64))
+			.saturating_add(T::DbWeight::get().reads(6722 as u64))
```
Comment (ggwpez) [inline frame/staking/src/weights.rs:406]: I am just seeing this now :sweat_smile:  
It is either a bug (fixed in https://github.com/paritytech/substrate/pull/12482) or expected.
